### ISSUE TITLE
How to get autocomplete for node.js modules?

### ISSUE THREAD

**Issue by ishanatmuz** [adobe/brackets#11929]
Hello All

Is there a way to get autocomplete for node.js modules inside brackets ?
It would also be very helpful to get autocomplete for modules that I am installing using npm.

If there is no current way but there is a possible way to make an extension for the same, just point me in the right direction and I will happily try to make an extension for this.

**Comment by ficristo** [comment]
Brackets use [`tern`](http://ternjs.net/) to get the autocompletion. But ATM you cannot configure the plugin, so is a bit limited.
You could try https://github.com/MiguelCastillo/Brackets-Ternific extension.
With it you should be able to create a `.tern-project` file and configure it to have better node autocomplete.

**Comment by yacut** [comment]
Can me someone explain: Why the nodejs support is not a priority for Brackets?

Today developers uses javascript for frontend and **for backend**, but code hints only works for the frontend project. Why is that?
Ternjs supports nodejs. And I guess to enable the node plugin will not be difficult.

**Comment by fdbozzo** [comment]
I agree with yacut. Hope better nodejs support.

**Comment by redmunds** [comment]
Search on DEFAULT_EXCLUDED_DIRECTORIES (in extensions/default/JavaScriptCodeHInts) and you'll see that the node_modules folder is excluded by default because it's generally very large and a big impact on performance. Try removing that and see how it works for you.

Also read through the [JS Code Hints wiki page](https://github.com/adobe/brackets/wiki/JavaScript-Code-Hints) for more info about further configuration options.

**Comment by fdbozzo** [comment]
Hi redmunds:

The preferences.js file in the directory you mention have the node_modules dir and require.* files excluded. I've commented them out and asigned null to them, but I do not have autocomplete for requires either.

The comments in the settings say "**by default exclude node_modules for performance reasons and because we don't do full hinting for those anyhow**":

```
        var BASE_EXCLUDED_DIRECTORIES = null, /* if the user has settings, we don't exclude anything by default */
            // exclude node_modules for performance reasons and because we don't do full hinting for those anyhow.
            //DEFAULT_EXCLUDED_DIRECTORIES = /node_modules/,
            DEFAULT_EXCLUDED_DIRECTORIES = null,
            // exclude require and jquery since we have special knowledge of those
            //BASE_EXCLUDED_FILES = /^require.*\.js$|^jquery.*\.js$/,
            BASE_EXCLUDED_FILES = null,

```
I've restarted Brackets, but autocompletion for requires does not work either :(

I'll keep reading...

**Comment by redmunds** [comment]
Not sure. Have you used any of the other "exclude" type extensions to speed up other parts of Brackets (such as as Find in Files) ? If so, excluding node_modules folder could also affect this (just a guess).

**Comment by fdbozzo** [comment]
Yes, I've used one, but I've disabled it before testing this again.
Thanks for your interest.

**Comment by pkelling** [comment]
I found a solution if anyone is still looking for hints/auto-complete for node.js in Brackets.

![screen shot 2017-03-09 at 2 59 41 pm](https://cloud.githubusercontent.com/assets/11371812/23770609/4eef0280-04d9-11e7-9ec4-2067e0e0fd7a.png)

Tern already includes a plugin for node.js, it just isn't being activated (and the angular.js one is). Type `angular.` and you should see hints for it pop up.

The line numbers might only be accurate for Release 1.9:
You'll need to open tern-worker.js (brackets/src/extensions/default/JavaScriptCodeHints)
-it's best if you're in hack mode: https://github.com/adobe/brackets/wiki/How-to-Hack-on-Brackets

1) on line 39: add to array -> `"tern/plugin/node"`

2) Then, go to line 136 and add `node: true` to the plugins {object}

![screen shot 2017-03-09 at 3 00 21 pm](https://cloud.githubusercontent.com/assets/11371812/23770617/5b14d5d0-04d9-11e7-9af0-35826440f178.png)


A couple side notes:
You can disable angular if you don't use it, by deleting `angular: true` or making it false.

The hints for js(ecmascript) and jquery (and others) are stored in (brackets/src/extensions/default/JavaScriptCodeHints/defs/).

The hints for node.js (and others) are stored in (brackets/src/extensions/default/JavaScriptCodeHints/plugin/) (node.js)

**Comment by ZacBouh** [comment]
Hello, I have tried the different solutions suggested here (and googled as much as I could) to try to solve the problem but I still dont get brackets to provide autocomplete for Node.JS modules.
Is there a more up to date solution  ?
